### PR TITLE
Exclude errors generated from A4A iframe

### PR DESCRIPTION
**Pull request by tiendt**
User error Reporting:
Filter out errors generated from A4A iframe embed

### CODE REVIEW

**Comment by tiendt** [comment]
This is just initial skeleton, I'm hoping to see whether the if else logic in user() is good. The next thing to do is to get ampdoc here (right now there's no win object) @lannka

File: src/log.js [hunk 1]
```diff
@@ -535,22 +536,34 @@ export function resetLogConstructorForTesting() {
  *  2. Development mode is enabled via `#development=1` or logging is explicitly
  *     enabled via `#log=D` where D >= 1.
  *
+ * @param {!Element=} opt_element
  * @return {!Log}
  */
-export function user() {
+export function user(opt_element) {
   if (logs.user) {
```
Comment (lannka) [inline src/log.js]: we should now cache 2 instances of user logger.

```
logs.user
logs.userForEmbed
```

logic will be like:

```javascript
if (isFromEmbed(opt_element) {
  if (!logs.userForEmbed) {
     logs.userForEmbed = ...
  }
  return logs.userForEmbed;
} else {
  if (!logs.user) {
    logs.user = ...
  }
  return logs.user;
}

```

File: src/log.js [hunk 2]
```diff
@@ -34,12 +34,27 @@ export const USER_ERROR_SENTINEL = '\u200B\u200B\u200B';
...
  * @return {boolean} Whether this message was a user error.
  */
 export function isUserErrorMessage(message) {
   return message.indexOf(USER_ERROR_SENTINEL) >= 0;
 }
 
+/**
+ * @return {boolean} Whether this message was a iframe embed error.
+ */
+function isErrorEmbed(message) {
```
Comment (lannka) [inline src/log.js]: isUserErrorEmbed

File: src/log.js [hunk 3]
```diff
@@ -34,12 +34,27 @@ export const USER_ERROR_SENTINEL = '\u200B\u200B\u200B';
 
 
 /**
+ * Four zero width space.
+ *
+ * @const {string}
+ */
+export const EMBED = '\u200B\u200B\u200B\u200B';
```
Comment (lannka) [inline src/log.js]: USER_ERROR_EMBED_SENTINEL

File: src/log.js [hunk 4]
```diff
@@ -535,9 +555,35 @@ export function resetLogConstructorForTesting() {
...
-export function user() {
+export function user(opt_element) {
+  const logger = getUserLogger();
+
+  if (!!opt_element &&
+      isFromEmbed(logger.win, /** @type {!Element} */ (opt_element))) {
+    if (logs.userForEmbed) {
+      return logs.userForEmbed;
+    }
+    logs.userForEmbed = new logConstructor(self, mode => {
```
Comment (lannka) [inline src/log.js]: seems duplicated logic, can we consolidate with getUserLogger?

File: src/error.js [hunk 5]
```diff
@@ -97,8 +97,10 @@ let detectedJsEngine;
  */
 export function reportErrorForWin(win, error, opt_associatedElement) {
   reportError(error, opt_associatedElement);
-  if (error && isUserErrorMessage(error.message) && !!win) {
-    reportErrorToAnalytics(/** @type {!Error} */(error), win);
+  if (error && !!win) {
+    if (isUserErrorMessage(error.message) && !error.embed) {
```
Comment (lannka) [inline src/error.js]: seems no nested if needed

File: test/functional/test-log.js [hunk 6]
```diff
@@ -673,5 +674,33 @@ describe('Logging', () => {
...
+      document.body.appendChild(iframe);
+    });
+
+    afterEach(() => {
+      sandbox.restore();
+      document.body.removeChild(iframe);
+    });
+
+    it('should return logger for user-error', () => {
+      expect(user().suffix_).to.equal(USER_ERROR_SENTINEL);
```
Comment (lannka) [inline test/functional/test-log.js]: you can use 
```
const error = user().createError();
expect(isUserErrorEmbed(error.message)).to.be.false;
```

also add test to make sure you not not creating extra identical logger

```
expect(user()).to.equal(user(element1));
expect(user()).to.equal(user(element2));
```

File: build-system/tasks/presubmit-checks.js [hunk 7]
```diff
@@ -341,6 +341,7 @@ var forbiddenTerms = {
       'extensions/amp-access/0.1/amp-access.js',
       'extensions/amp-experiment/0.1/variant.js',
       'extensions/amp-user-notification/0.1/amp-user-notification.js',
+      'dist.3p/current/integration.js',
```
Comment (zhouyx) [inline build-system/tasks/presubmit-checks.js]: what's this for?

File: src/log.js [hunk 8]
```diff
@@ -210,6 +229,7 @@ export class Log {
     const error = this.error_.apply(this, arguments);
     if (error) {
       error.name = tag || error.name;
+      error.embed = isUserErrorEmbed(error.message);
```
Comment (tiendt) [inline src/log.js]: yeah i just realized, seems like error.embed is not needed anymore

File: src/log.js [hunk 9]
```diff
@@ -34,12 +34,27 @@ export const USER_ERROR_SENTINEL = '\u200B\u200B\u200B';
...
+
+/**
  * @return {boolean} Whether this message was a user error.
  */
 export function isUserErrorMessage(message) {
   return message.indexOf(USER_ERROR_SENTINEL) >= 0;
 }
 
+/**
+ * @return {boolean} Whether this message was a iframe embed error.
```
Comment (lannka) [inline src/log.js]: Whether this message was a a user error from an iframe embed

File: src/log.js [hunk 10]
```diff
@@ -540,25 +559,45 @@ export function resetLogConstructorForTesting() {
  *  2. Development mode is enabled via `#development=1` or logging is explicitly
  *     enabled via `#log=D` where D >= 1.
  *
+ * @param {!Element=} opt_element
  * @return {!Log}
  */
-export function user() {
+export function user(opt_element) {
+  let logger;
```
Comment (lannka) [inline src/log.js]: seems this local var is not needed

File: test/functional/test-log.js [hunk 11]
```diff
@@ -673,5 +675,45 @@ describe('Logging', () => {
...
+
+    afterEach(() => {
+      sandbox.restore();
+      document.body.removeChild(iframe);
+    });
+
+    it('should return logger for user-error', () => {
+      const error = user().createError();
+      expect(isUserErrorEmbed(error.message)).to.be.false;
+      expect(user(this.element).suffix_).to.equal(USER_ERROR_SENTINEL);
```
Comment (lannka) [inline test/functional/test-log.js]: avoid checking member field

```
expect(isUserError(error.message)).to.be.true;
```

File: test/functional/test-log.js [hunk 12]
```diff
@@ -673,5 +675,45 @@ describe('Logging', () => {
...
+    });
+
+    it('should return logger for embed-error', () => {
+      element = document.createElement('embed');
+      iframe.contentWindow.document.body.appendChild(element);
+      expect(user(element).suffix_).to.equal(USER_ERROR_EMBED_SENTINEL);
```
Comment (lannka) [inline test/functional/test-log.js]: same here

File: test/functional/test-log.js [hunk 13]
```diff
@@ -673,5 +675,45 @@ describe('Logging', () => {
...
+    });
+
+    it('should not create extra identical logs', () => {
```
Comment (lannka) [inline test/functional/test-log.js]: loggers

File: test/functional/test-log.js [hunk 14]
```diff
@@ -673,5 +675,45 @@ describe('Logging', () => {
...
+      element1 = document.createElement('embed_1');
+      element2 = document.createElement('embed_2');
+      iframe.contentWindow.document.body.appendChild(element1);
+      iframe.contentWindow.document.body.appendChild(element2);
+      expect(user()).to.equal(user(this.element));
+      expect(user(element1)).to.equal(user(element2));
```
Comment (lannka) [inline test/functional/test-log.js]: add one more check:
```
expect(user()).to.not.equal(user(element1));
```

File: src/log.js (hunk 10 above)
Comment (tiendt) [inline src/log.js]: @lannka i think it is because I'd need to assign different value to this var and also use it at the last return call. Lmk what you think!

File: build-system/tasks/presubmit-checks.js (hunk 7 above)
Comment (tiendt) [inline build-system/tasks/presubmit-checks.js]: @zhouyx When i run presubmit it shows that i'm using cid, so Hongfei suggested to whitelist it.

File: src/log.js (hunk 10 above)
Comment (lannka) [inline src/log.js]: why don't you just assign to `logs.user` ?

File: build-system/tasks/presubmit-checks.js (hunk 7 above)
Comment (lannka) [inline build-system/tasks/presubmit-checks.js]: sorry if there's mis-understanding, but I don't see it cid used in your change. did you try make a new build and run presubmit again?

File: test/functional/test-log.js (hunk 14 above)
Comment (tiendt) [inline test/functional/test-log.js]: @lannka i made all requested change. Thanks!

File: src/log.js [hunk 15]
```diff
@@ -540,25 +559,45 @@ export function resetLogConstructorForTesting() {
...
  *     enabled via `#log=D` where D >= 1.
  *
+ * @param {!Element=} opt_element
  * @return {!Log}
  */
-export function user() {
+export function user(opt_element) {
+  //let logger;
   if (logs.user) {
+    logs.user = logs.user;
```
Comment (lannka) [inline src/log.js]: why this?

File: builtins/amp-img.js [hunk 16]
```diff
@@ -123,9 +123,10 @@ export class AmpImg extends BaseElement {
     // only read "Graphic" when using only 'alt'.
     if (this.element.getAttribute('role') == 'img') {
       this.element.removeAttribute('role');
-      user().error('AMP-IMG', 'Setting role=img on amp-img elements breaks ' +
-        'screen readers please just set alt or ARIA attributes, they will ' +
-        'be correctly propagated for the underlying <img> element.');
+      user(this.element).error('AMP-IMG',
```
Comment (zhouyx) [inline builtins/amp-img.js]: Question: Do we plan to migrate the new `user(opt_element)` in this PR?  If so why we only make the change to `amp-img.js`. I would suggest migrate to new format in separate PR.

Comment (tiendt) [inline builtins/amp-img.js]: @zhouyx I need that change to amp-img.js to do the test, but the whole migration to use(opt_element) would be in a separate PR. Is that ok?

File: src/log.js [hunk 17]
```diff
@@ -587,3 +622,15 @@ export function dev() {
...
 }
+
+/**
+ * @param {!Window} win
+ * @param {!Element=} opt_element
+ * @returns {boolean} isEmbed
+ */
+export function isFromEmbed(win, opt_element) {
+  if (!!opt_element) {
+    return opt_element.ownerDocument.defaultView != win;
```
Comment (lannka) [inline src/log.js]: nitpick: logically cleaner to flip the `if`

**Comment by lannka** [comment]
LGTM @zhouyx pls help merge once you have done the review.

File: builtins/amp-img.js (hunk 16 above)
Comment (zhouyx) [inline builtins/amp-img.js]: Talked offline. Please revert. and I'll approve and merge the PR.

Comment (tiendt) [inline builtins/amp-img.js]: @zhouyx reverted!
